LiteCore/Storage/SQLiteDataFile.cc

```diff
@@ -499,11 +541,45 @@ path.path().c_str());
         }
     }
 
+    void SQLiteDataFile::collectBlobs(string storeName, unordered_set<string> &digests) {
+        stringstream s;
+        s << "SELECT body FROM kv_" << storeName << " WHERE (flags & " <<
+            (int)DocumentFlags::kHasAttachments << ") != 0 AND (flags & " <<
+            (int)DocumentFlags::kDeleted << ") == 0";
+        
+        string sql = s.str();
+        compile(_collectBlobStmt, sql.c_str());
+        UsingStatement u(_collectBlobStmt);
+        while(_collectBlobStmt->executeStep()) {
+            slice sl = SQLiteKeyStore::columnAsSlice(_collectBlobStmt->getColumn(0));
+            slice body = RawRevision::getCurrentRevBody(sl);
+            const Dict* root = Value::fromTrustedData(body)->asDict();
+            findBlobs(root, digests, documentKeys());
+        }
+    }
+    
+    void SQLiteDataFile::removeUnusedBlobs(const unordered_set<string> &used) {
+        FilePath attsPath(filePath().dirName() + "Attachments", "");
+        attsPath.forEachFile([&used](const FilePath &path) {
+            string digest = string("sha1-") + path.fileName().substr(0, path.fileName().length() - 5);
+            replace(digest.begin(), digest.end(), '_', '/');
+            if(find(used.cbegin(), used.cend(), digest) == used.cend()) {
+                path.del();
+            }
+        });
+    }
 
     void SQLiteDataFile::compact() {
         checkOpen();
         beganCompacting();
         maybeVacuum();
+        auto storeNames = allKeyStoreNames();
+        unordered_set<string> digestsInUse;
+        for_each(storeNames.cbegin(), storeNames.cend(), [&](const string &storeName) {
```

Comment:
Only the default KeyStore contains documents. The others just have key/value stuff like checkpoints and other db metadata. DataFile itself doesn't know about which ones contain what, so this logic probably belongs up at the Database class.